Generate access list allowed promotions on access request creation

jakule opened the pull request:
Access request suggestion API returns valid access list promotions with the same or more access and can replace/promote the current request. The initial API was introduced here https://github.com/gravitational/teleport.e/pull/2084

This PR solves two main issues that were reported after testing:
1. as the suggestions are computed at the query time reviewer context is used to fetch all required resources. The reviewer may not have access to them, resulting in the API returning an error.
2. the suggestions are computed on every query, which can be very costly.

This change moves the computation process to access request creation and stores the promotions in the DB so they can be cheaply retrieved when needed. Because the computation will be using requester context now the problem with lacking access to resources will also be solved.

jakule committed:
Generate access list suggestions on access request creation

jakule committed:
Add test
Cleanup code

jakule committed:
Add comment

jakule committed:
Fix typo and refactor access suggestion in services library

Fixed a typo in a comment in `access_list.go`. Refactored `accessrequest/suggestion.go` for enhanced code structure. The old model had all methods related to access request in a single interface, `accessRequestAPIGetter`, which had been divided among multiple interfaces for better division of responsibilities. A new function `GetSuggestedAccessLists` was created to narrow down the single duty for each function. These changes ensure more convenient unit testing and better accountability.

jakule committed:
Improve access request suggestions logic

This commit improves the logic for generating access request suggestions. It includes the requester's roles and traits as well as those from the access list and provides suggestions only for resource-based requests. A new method was also introduced for initializing AccessRequestSuggestions to avoid nil error.

xacrimon reviewed (commented):
do we want some godocs on the various interfaces and exported methods on various services etc as is pattern?

jakule committed:
Refactor access list suggestions to promotions

The concept of 'suggestions' was updated to 'promotions' in different files (auth.go, client.go, auth_service_test.go, auth_service.proto among others) to make it more intuitive and align with the business context. This implies changing the wording, the functions called and the variable names. The feature itself hasn't changed, just the way it is referred.

jakule committed:
This commit changes the method name "GenerateAccessListSuggestions" to "GenerateAccessRequestPromotions" in multiple files. The naming is changed to reflect the actual function of the method, which is generating promotions for given access requests, rather than generating list suggestions. All relevant mentions and usages of this method are updated to reflect the name change. Modification also includes adjustments to the comments and interfaces related to the updated function for better code clarity.

jakule committed:
Refactor "GenerateAccessListSuggestions" to "GenerateAccessRequestPromotions"

Renamed method "GenerateAccessListSuggestions" to "GenerateAccessRequestPromotions" across the codebase. This change reflects the actual functionality of the method, which is to generate promotions for access requests, not suggestions. This aligns the method name with its purpose, and offers better code readability. Comments related to this function were also updated to maintain consistency.

jakule committed:
Update method handling access requests promotions

Revised the method of handling failures when generating access requests promotions. Now, even if promotions fail to generate, the request will not fail entirely. Instead, the request can still be approved but without promotions. Also, changed the UpsertAccessRequestAllowedPromotions method to CreateAccessRequestAllowedPromotions to clearly indicate its purpose, while revising related test cases and comments.

fspmarshall commented on lib/auth/auth.go:
```diff
@@ -4174,6 +4174,20 @@ func (a *Server) CreateAccessRequestV2(ctx context.Context, req types.AccessRequ
+		// Create the promotion entry even if the allowed promotion is empty. Otherwise, we won't
+		// be able to distinguish between an allowed empty set and generation failure.
+		if err := a.Services.CreateAccessRequestAllowedPromotions(ctx, reqCopy, promotions); err != nil {
+			log.WithError(err).Warn("Failed to update access request with promotions.")
```
nit: I'd clarify this a bit so it doesn't read like the access request itself is blocked by the error.
```suggestion
			log.WithError(err).Warn("Failed to store allowed access request promotions.")
```
